File: solutions/cpp/rna-transcription/3/rna_transcription.cpp
#include "rna_transcription.h"

#include <map>

namespace rna_transcription {
std::map<char, char> transcription_pairs = {
    {'C', 'G'}, {'G', 'C'}, {'A', 'U'}, {'T', 'A'}};

void validate_dna_nucleotide(char nucleotide) {
	if (nucleotide != 'A' && nucleotide != 'C' && nucleotide != 'G' &&
	    nucleotide != 'T') {
		throw std::invalid_argument{"Character is not a valid nucleotide"};
	};
}

char to_rna(char nucleotide) {
	validate_dna_nucleotide(nucleotide);
	return transcription_pairs.at(nucleotide);
}

std::string to_rna(std::string dna) {
	std::string rna = "";
	for (char nucleotide : dna) {
		rna += to_rna(nucleotide);
	}

	return rna;
}

}  // namespace rna_transcription

File: solutions/fortran/rna-transcription/3/rna_transcription.f90
module rna_transcription
    implicit none
contains

function to_rna(dna) result(rna)
    character(*), intent(in) :: dna
    character(len(dna)) :: rna
    integer :: i
      
    do i = 1, len(dna)
        select case(dna(i:i))
        case('G')
            rna(i:i) = 'C'
        case('C')
            rna(i:i) = 'G'
        case('T')
            rna(i:i) = 'A'
        case('A')
            rna(i:i) = 'U'
        case default
            print *, "Invalid DNA base:", dna(i:i), "at position", i
            error stop 1
        end select
    end do
end function to_rna

end module rna_transcription
